### PR TITLE
Fix bug in notification generation logic

### PR DESCRIPTION
**Pull request by jasonodoom**
#  #

## 🗣 Description ##

This PR fixes a logic bug that was preventing some parent organizations' notification PDFs from being generated.

It also updates some changes to improve logging (see https://github.com/cisagov/cyhy-reports/pull/104/commits/da0378fc2a27967a3166627b7813793f78713826 and https://github.com/cisagov/cyhy-reports/pull/104/commits/d9c9891d3ebd09ea6076f65a0155a3a3a0949076).

## 💭 Motivation and context ##

Today's daily notification generation process did not generate PDFs for certain organizations that should have received them. Specifically, organizations that do receive notification PDFs and that have children that do not. This PR fixes that issue.

The logging change in https://github.com/cisagov/cyhy-reports/pull/104/commits/d9c9891d3ebd09ea6076f65a0155a3a3a0949076 was done because [this log message](https://github.com/cisagov/cyhy-reports/blob/d9c9891d3ebd09ea6076f65a0155a3a3a0949076/extras/create_send_notifications.py#L209-L213) was listing orgs that had notifications included in an ancestor's notification PDF, which was misleading and confusing.

## 🧪 Testing ##

The cyhy-reports docker image was rebuilt and run locally with these changes and I was able to verify that the missing parent organization notification PDFs that were not present previously now get created. 

This change was sent to Production and is confirmed to also be working. 

## ✅ Pre-approval checklist ##

- [X] This PR has an informative and human-readable title.
- [X] Changes are limited to a single goal - *eschew scope creep!*
- [X] All relevant type-of-change labels have been added.
- [X] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [X] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

- [x] Deploy this change to Production.

### CODE REVIEW

File: extras/create_send_notifications.py [hunk 1]
```diff
@@ -51,14 +51,16 @@ def build_notifications_org_list(db):
...
     cyhy_parent_ids = set()
     ticket_owner_ids = db.NotificationDoc.collection.distinct("ticket_owner")
-    for request in db.RequestDoc.collection.find({"_id": {"$in": ticket_owner_ids}, "report_types": "CYHY"}, {"_id":1}):
-        # If the notification document's ticket owner has "CYHY" in their list of report_types,
-        # then a notification should be generated for that owner:
-        notifications_to_generate.add(request["_id"])
-        logging.debug("Added {} to notifications_to_generate".format(request["_id"]))
+    for request in db.RequestDoc.collection.find({"_id": {"$in": ticket_owner_ids}}, {"_id": 1, "report_types": 1}):
+        if "CYHY" in request["report_types"]:
+            # If the notification document's ticket owner has "CYHY" in their list of report_types,
```
Comment (jsf9k) [inline extras/create_send_notifications.py]:
```suggestion
            # If the notification document's ticket owner has "CYHY" in its list of report_types,
```

**Review by jsf9k — CHANGES_REQUESTED** [review]
Requesting one minor wording change...

**Review by jsf9k — APPROVED** [review]

**Comment by jsf9k** [comment]
The "Motivation" section of the PR description and the commit message for b42697d are unclear.  You say, "...we excluded those notification ticket owners from being called..." but I feel like a phrase is missing; something like "...we excluded those notifications ticket owners _that do not have CyHy in their report types from...".  The "being called" that follows immediately afterward is also unclear.  What do you mean by "being called"?

**Review by dav3r — APPROVED** [review]
LGTM! 🚫 🐞

**Review by jsf9k — CHANGES_REQUESTED** [review]
The commit message still needs to be redone. You can just use what you wrote in the "Motivation" section.

File: extras/create_send_notifications.py [hunk 2]
```diff
@@ -51,14 +51,16 @@ def build_notifications_org_list(db):
...
+    for request in db.RequestDoc.collection.find({"_id": {"$in": ticket_owner_ids}}, {"_id": 1, "report_types": 1}):
+        if "CYHY" in request["report_types"]:
+            # If the notification document's ticket owner has "CYHY" in its list of report_types,
+            # then a notification should be generated for that owner:
+            notifications_to_generate.add(request["_id"])
+            logging.debug("Added {} to notifications_to_generate".format(request["_id"]))
         # Recursively check for any ancestors of the ticket owner that have "CYHY" in
         # their list of report_types.  If found, add them to the list of owners that
         # should get a notification.
+        logging.debug("{} doesn't have CYHY in report_types; checking parents, grandparents, etc.".format(request["_id"]))
```
Comment (dav3r) [inline extras/create_send_notifications.py]: After further thought, this is a better and more correct way to state what is happening here:
```suggestion
        logging.debug("Checking for ancestors of {} with CYHY in their list of report_types".format(request["_id"]))
```

File: extras/create_send_notifications.py [hunk 3]
```diff
@@ -81,6 +83,7 @@ def find_cyhy_parents(db, org_id):
             logging.debug("{} - Adding to set of CYHY parents".format(request["_id"]))
         # Recursively call find_cyhy_parents() to check if this org has any parents
         # with "CYHY" in their list of report_types
+        logging.debug("{} doesn't have CYHY in report_types; checking parents, grandparents, etc.".format(request["_id"]))
```
Comment (dav3r) [inline extras/create_send_notifications.py]: Same as above:
```suggestion
        logging.debug("Checking for ancestors of {} with CYHY in their list of report_types".format(request["_id"]))
```

**Comment by jasonodoom** [comment]
> The commit message still needs to be redone. You can just use what you wrote in the "Motivation" section.

@jsf9k  Please see https://github.com/cisagov/cyhy-reports/pull/104/commits/cdc3fe10facff3716fed5848fe1d0896842d65a7

**Comment by jsf9k** [comment]
> > The commit message still needs to be redone. You can just use what you wrote in the "Motivation" section.
> 
> @jsf9k  Please see https://github.com/cisagov/cyhy-reports/pull/104/commits/cdc3fe10facff3716fed5848fe1d0896842d65a7

That commit message is also unclear. Please just use the first paragraph of the "Motivation" section of the PR description.

**Comment by jasonodoom** [comment]
> > > The commit message still needs to be redone. You can just use what you wrote in the "Motivation" section.
> > 
> > 
> > @jsf9k  Please see [cdc3fe1](https://github.com/cisagov/cyhy-reports/commit/cdc3fe10facff3716fed5848fe1d0896842d65a7)
> 
> That commit message is also unclear. Please just use the first paragraph of the "Motivation" section of the PR description.

@jsf9k  Please see https://github.com/cisagov/cyhy-reports/pull/104/commits/655711ab00d73dc68458d35805dab50cb057c086

The first line should be a quick summary of the change made, not the first paragraph of the in-depth description.  I _beg_ you to read [this](https://cbea.ms/git-commit/), then add a short summarizing line at the very beginning of your commit message.  Everything you have above should stay in, but you should add a single line at the very top summarizing the changes that were made.

**Review by jsf9k — APPROVED** [review]

**Comment by jsf9k** [comment]
That's much better @jasonodoom.  Thank you!

**Review by dav3r — APPROVED** [review]
Approval intensifies! ⚡

File: extras/create_send_notifications.py [hunk 4]
```diff
@@ -51,18 +51,20 @@ def build_notifications_org_list(db):
...
         # their list of report_types.  If found, add them to the list of owners that
         # should get a notification.
+        logging.debug("Checking for ancestors of {} with CYHY in their list of report_types".format(request["_id"]))
         cyhy_parent_ids.update(find_cyhy_parents(db, request["_id"]))
     notifications_to_generate.update(cyhy_parent_ids)
     notifications_to_delete = set(ticket_owner_ids) - notifications_to_generate
-    return list(notifications_to_generate), list(notifications_to_delete)
+    return sorted(notifications_to_generate), list(notifications_to_delete)
           
 def find_cyhy_parents(db, org_id):
```
Comment (mcdonnnj) [inline extras/create_send_notifications.py:68]: I'm not sure if we've noted this anywhere but the way this function works if there is a mistake in stakeholder configurations such that a cycle is created (A is an ancestor of B and B is an ancestor of A) I believe this will recursively overflow.

File: extras/create_send_notifications.py [hunk 5]
```diff
@@ -188,6 +196,11 @@ def main():
     else:
         logging.info("Nothing to email - skipping this step")
 
+    # Remove orgs from notifications_to_delete if they are in the list of orgs
+    # that we just generated notifications for (most likely because the
+    # notification was included in an ancestor org's notification)
+    notifications_to_delete = sorted(set(notifications_to_delete) - set(orgs_notified))
```
Comment (mcdonnnj) [inline extras/create_send_notifications.py]: Thoughts on changing the first use of `notifications_to_delete` to something like `notifications_not_generated` (and propagating this proposed name change to the same variable name in the `build_notifications_org_list()` function)? I'm generally not a fan of overwriting variables with derived values as it makes the original variable name slightly misleading.

**Review by mcdonnnj — APPROVED** [review]
This looks pretty solid. Thank you both for working on this 💪. I have one comment/note and one suggestion for consideration.

File: extras/create_send_notifications.py (hunk 4 above)
Comment (dav3r) [reply extras/create_send_notifications.py:68]: We do [check for cycles in our cyhy-suborg tool](https://github.com/cisagov/cyhy-core/blob/9d867f753136747e709480c770cc7021c9253849/bin/cyhy-suborg#L158-L164) (which is the "official" way to manage CyHy sub-orgs), so I think that's sufficient to mitigate your concern here.

File: extras/create_send_notifications.py (hunk 5 above)
Comment (jsf9k) [reply extras/create_send_notifications.py]: I like it!

File: extras/create_send_notifications.py (hunk 4 above)
Comment (jsf9k) [reply extras/create_send_notifications.py:68]: Would a short comment pointing out the potential pitfall be appropriate here?

Comment (dav3r) [reply extras/create_send_notifications.py:68]: It couldn't hurt to add such a comment, something along the lines of:
```
# WARNING: Even though cyhy-suborg explicitly checks for and blocks stakeholder
# cycles, if that guardrail is ever subverted, this function will recursively
# overflow and fail.
```

Comment (mcdonnnj) [reply extras/create_send_notifications.py:68]: Ah, right. Thank you for the reminder! I think that comment is perfectly suitable.

Comment (jasonodoom) [reply extras/create_send_notifications.py:68]: Please see https://github.com/cisagov/cyhy-reports/pull/104/commits/6c0aa802e2f184ce0136e13b0168688361557add

File: extras/create_send_notifications.py (hunk 5 above)
Comment (jasonodoom) [reply extras/create_send_notifications.py]: Please see https://github.com/cisagov/cyhy-reports/pull/104/commits/2e1377549d690b7d7899661bf9e01ea4eb2d2d54

File: extras/create_send_notifications.py [hunk 6]
```diff
@@ -188,6 +201,11 @@ def main():
     else:
         logging.info("Nothing to email - skipping this step")
 
+    # Remove orgs from notifications_to_delete if they are in the list of orgs
+    # that we just generated notifications for (most likely because the
+    # notification was included in an ancestor org's notification)
+    notifications_not_generated = sorted(set(notifications_to_delete) - set(orgs_notified))
```
Comment (mcdonnnj) [inline extras/create_send_notifications.py]: Apologies that this was unclear in https://github.com/cisagov/cyhy-reports/pull/104#discussion_r1416154995:
```suggestion
    notifications_to_delete = sorted(set(notifications_not_generated) - set(orgs_notified))
```

and then `notifications_not_generated` instead of `notifications_to_delete` in
https://github.com/cisagov/cyhy-reports/blob/2e1377549d690b7d7899661bf9e01ea4eb2d2d54/extras/create_send_notifications.py#L141

**Review by mcdonnnj — CHANGES_REQUESTED** [review]

File: extras/create_send_notifications.py [hunk 7]
```diff
@@ -51,22 +51,29 @@ def build_notifications_org_list(db):
...
         # Recursively check for any ancestors of the ticket owner that have "CYHY" in
         # their list of report_types.  If found, add them to the list of owners that
         # should get a notification.
+        logging.debug("Checking for ancestors of {} with CYHY in their list of report_types".format(request["_id"]))
         cyhy_parent_ids.update(find_cyhy_parents(db, request["_id"]))
     notifications_to_generate.update(cyhy_parent_ids)
-    notifications_to_delete = set(ticket_owner_ids) - notifications_to_generate
-    return list(notifications_to_generate), list(notifications_to_delete)
+    notifications_not_generated = set(ticket_owner_ids) - notifications_to_generate
+    return sorted(notifications_to_generate), list(notifications_not_generated)
```
Comment (jsf9k) [inline extras/create_send_notifications.py:66]: "Return notifications organizations list" does not really describe the two `list`s being returned by this function.  Can you update the pydoc text to describe what the two `list`s being returned actually are?

File: extras/create_send_notifications.py
Comment (jsf9k) [inline extras/create_send_notifications.py:1]: I think
```python
notifications_org_ids, notifications_to_delete = build_notifications_org_list(db)
```
on line 141 of this file should actually be
```python
notifications_org_ids, notifications_not_generated = build_notifications_org_list(db)
```

Otherwise I think `notifications_not_generated` is undefined on line 207.  In any event, after the variable rename this code needs to be tested again.

**Review by jsf9k — CHANGES_REQUESTED** [review]

File: extras/create_send_notifications.py [hunk 8]
```diff
@@ -51,22 +51,29 @@ def build_notifications_org_list(db):
...
           
 def find_cyhy_parents(db, org_id):
     """Return parents/grandparents/etc. of an organization that have "CYHY" in their list of report_types.
     """
+
+    # WARNING: Even though cyhy-suborg explicitly checks for and blocks stakeholder
+    # cycles, if that guardrail is ever subverted, this function will recursively
+    # overflow and fail.
+
```
Comment (dav3r) [inline extras/create_send_notifications.py]: I know I wrote this comment, but looking at it with fresh eyes, it makes sense to define what we mean by "stakeholder cycles":
```suggestion
    # WARNING: Even though cyhy-suborg explicitly checks for and blocks
    # stakeholder cycles (e.g. org A is a descendant of org B, which is a
    # descendant of org A), if that guardrail is ever subverted, this function
    # will recursively overflow and fail.
```

File: extras/create_send_notifications.py (hunk 7 above)
Comment (jasonodoom) [reply extras/create_send_notifications.py:66]: Please see https://github.com/cisagov/cyhy-reports/pull/104/commits/65e11a23ccb2261b5db7673640dfc5e246e31c5c

File: extras/create_send_notifications.py [hunk 9]
```diff
@@ -43,7 +43,8 @@ def create_output_directories():
     )
 
 def build_notifications_org_list(db):
-    """Return notifications organization list.
+    """Return list of notifications to generate and 
+       notifications not generated.
```
Comment (jsf9k) [inline extras/create_send_notifications.py]: I think this is clearer and more correct since the lists contain org IDs, not notifications:
```suggestion
    """Return a list of org IDs for which notifications should be
    generated as well as a list of org IDs for which notifications
    should not be generated.
```

It's a bit wordy, though, so maybe someone else can think of something more succinct.

File: extras/create_send_notifications.py [hunk 10]
```diff
@@ -43,7 +43,8 @@ def create_output_directories():
...
 
     This is the list of organization IDs that should
     have a notification generated and sent.
```
Comment (jsf9k) [inline extras/create_send_notifications.py]: This text should be removed or updated to be correct.

**Review by jsf9k — CHANGES_REQUESTED** [review]

File: extras/create_send_notifications.py (hunk 10 above)
Comment (jasonodoom) [reply extras/create_send_notifications.py]: Please see https://github.com/cisagov/cyhy-reports/pull/104/commits/b98cfd2f9a01aa2f98b6440a57e2de8f9a87c2a3

File: extras/create_send_notifications.py
Comment (jasonodoom) [reply extras/create_send_notifications.py:1]: @jsf9k Can confirm this is working after testing.

Comment (jsf9k) [reply extras/create_send_notifications.py:1]: See commit ee5fdc52520b51ae1c6878083494fa03bfdba1ba.

File: extras/create_send_notifications.py (hunk 7 above)
Comment (jsf9k) [reply extras/create_send_notifications.py:66]: See also commit 660c895.

File: extras/create_send_notifications.py (hunk 5 above)
Comment (jsf9k) [reply extras/create_send_notifications.py]: See also commit ee5fdc52520b51ae1c6878083494fa03bfdba1ba.

**Review by jsf9k — APPROVED** [review]
Thanks for sticking with the review process @jasonodoom!

File: extras/create_send_notifications.py (hunk 5 above)
Comment (mcdonnnj) [reply extras/create_send_notifications.py]: Also https://github.com/cisagov/cyhy-reports/pull/104/commits/7d844b3fcad1b3ea7d20f038ce203303e03a5a9c for completeness.

**Review by mcdonnnj — APPROVED** [review]
Thanks for sticking with us! LGTM ✔
